File: frontend/src/types/todo.ts
// User-related types
export interface User {
  id: string;
  name: string;
  email: string;
  created_at: string;
}

export interface UserCredentials {
  email: string;
  password: string;
}

export interface UserRegistration extends UserCredentials {
  name: string;
}

// Todo-related types
export interface Todo {
  id: string;
  user_id: string;
  title: string;
  description?: string;
  status: 'pending' | 'in-progress' | 'completed';
  created_at: string;
  updated_at: string;
}

export interface TodoCreate {
  title: string;
  description?: string;
  status?: 'pending' | 'in-progress' | 'completed';
}

export interface TodoUpdate {
  title?: string;
  description?: string;
  status?: 'pending' | 'in-progress' | 'completed';
}

// API response types
export interface ApiResponse<T> {
  data?: T;
  error?: string;
}

// Authentication response type
export interface AuthResponse {
  access_token: string;
  token_type: string;
  user: User;
}

File: backend/src/models/todo.py
from sqlmodel import SQLModel, Field
from datetime import datetime
from typing import Optional
import uuid
from pydantic import field_validator
from .user import User

class TodoBase(SQLModel):
    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=1000)
    status: str = Field(default="pending", regex="^(pending|in-progress|completed)$")

class Todo(TodoBase, table=True):
    id: Optional[uuid.UUID] = Field(default_factory=uuid.uuid4, primary_key=True)
    user_id: uuid.UUID = Field(foreign_key="user.id", nullable=False)
    created_at: Optional[datetime] = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = Field(default_factory=datetime.utcnow)

class TodoCreate(TodoBase):
    pass

class TodoUpdate(SQLModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=1000)
    status: Optional[str] = Field(default=None, regex="^(pending|in-progress|completed)$")

class TodoRead(TodoBase):
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime